src/front/routes/Items.tsx

```diff
@@ -35,34 +34,90 @@ const itemsQueryDocument = graphql(`
         }
     }
 `)
+const principles = [
+    "edge",
+    "forge",
+    "grail",
+    "heart",
+    "knock",
+    "lantern",
+    "moon",
+    "moth",
+    "nectar",
+    "rose",
+    "scale",
+    "sky",
+    "winter",
+] as const
 
 type ItemFromQuery = types.ItemsQuery["item"][number]
 
+interface ItemsProps {
+    filters?: {
+        known?: boolean
+        edge?: boolean
+        forge?: boolean
+        grail?: boolean
+        heart?: boolean
+        knock?: boolean
+        lantern?: boolean
+        moon?: boolean
+        moth?: boolean
+        nectar?: boolean
+        rose?: boolean
+        scale?: boolean
+        sky?: boolean
+        winter?: boolean
+        aspect?: string
+    }
+}
+type ItemsAction = ItemsActionInner | ItemsActionOuter
+type ItemsActionInner = never
+type ItemsActionOuter = { type: "filter"; filters: ItemsProps["filters"] }
+
+function filterItems(
+    filters: ItemsProps["filters"],
+    state: ItemFromQuery[],
+): ItemFromQuery[] {
+    let filtered_state = state
+    if (filters?.known) {
+        filtered_state = state.filter(({ known }) => known === true)
+    }
+    principles.forEach((principle) => {
```

Comment:
never `forEach`, use `const filteredState = ...` where `...` contains `map`, `flatMap`, `filter`, and/or `reduce` and if you really have to do things imperatively, use `for (const x of y)`